qa/restart-upgrade/src/test/java/org/opensearch/neuralsearch/bwc/HybridSearchIT.java

```diff
@@ -0,0 +1,163 @@
+    private static final String TEXT_1 = "Hello world";
+    private static final String TEXT_2 = "Hi planet";
+    private static final String TEXT_3 = "Hi earth";
+    private static final String TEXT_4 = "Hi amazon";
+    private static final String TEXT_5 = "Hi mars";
+    private static final String TEXT_6 = "Hi opensearch";
+    private static final String QUERY = "Hi world";
+
+    // Test restart-upgrade normalization processor when index with multiple shards
+    // Create Text Embedding Processor, Ingestion Pipeline, add document and search pipeline with normalization processor
+    // Validate process , pipeline and document count in restart-upgrade scenario
+    public void testNormalizationProcessor_whenIndexWithMultipleShards_E2EFlow() throws Exception {
+        waitForClusterHealthGreen(NODES_BWC_CLUSTER);
+        if (isRunningAgainstOldCluster()) {
+            String modelId = uploadTextEmbeddingModel();
+            loadModel(modelId);
+            createPipelineProcessor(modelId, PIPELINE_NAME);
+            createIndexWithConfiguration(
+                getIndexNameForTest(),
+                Files.readString(Path.of(classLoader.getResource("processor/IndexMappingMultipleShard.json").toURI())),
+                PIPELINE_NAME
+            );
+            addDocument(getIndexNameForTest(), "0", TEST_FIELD, TEXT_1, null, null);
+            addDocument(getIndexNameForTest(), "1", TEST_FIELD, TEXT_2, null, null);
+            addDocument(getIndexNameForTest(), "2", TEST_FIELD, TEXT_3, null, null);
+            addDocument(getIndexNameForTest(), "3", TEST_FIELD, TEXT_4, null, null);
+            addDocument(getIndexNameForTest(), "4", TEST_FIELD, TEXT_5, null, null);
+            createSearchPipeline(
+                SEARCH_PIPELINE_NAME,
+                DEFAULT_NORMALIZATION_METHOD,
+                DEFAULT_COMBINATION_METHOD,
+                Map.of(PARAM_NAME_WEIGHTS, Arrays.toString(new float[] { 0.3f, 0.7f }))
+            );
+        } else {
+            Map<String, Object> pipeline = getIngestionPipeline(PIPELINE_NAME);
+            assertNotNull(pipeline);
+            String modelId = getModelId(pipeline, TEXT_EMBEDDING_PROCESSOR);
+            loadModel(modelId);
+            addDocument(getIndexNameForTest(), "5", TEST_FIELD, TEXT_6, null, null);
+            validateTestIndex(modelId, getIndexNameForTest(), SEARCH_PIPELINE_NAME);
+            deleteSearchPipeline(SEARCH_PIPELINE_NAME);
+            deletePipeline(PIPELINE_NAME);
+            deleteModel(modelId);
+            deleteIndex(getIndexNameForTest());
+        }
+    }
+
+    // Test restart-upgrade normalization processor when index with single shard
+    // Create Text Embedding Processor, Ingestion Pipeline, add document and search pipeline with normalization processor
+    // Validate process , pipeline and document count in restart-upgrade scenario
+    public void testNormalizationProcessor_whenIndexWithSingleShard_E2EFlow() throws Exception {
```

Comment:
Only shard count is different from above test case. Can we remove code duplication?

Reply:
@martin-gaievski asked specifically to write this test case.

Reply:
@heemin32 we need a one shard case as it triggers different code flow mainly on core side, main difference is that execution for fetch and query phases is not guaranteed due to some optimizations. We built special handling for 1 shard case in [hybrid query code](https://github.com/opensearch-project/neural-search/blob/main/src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java#L131-L133).

Reply:
if concern is more around code duplication, I agree we should spent some effort to get more optimal code and organize it in a way that code is re-used and just number of shards/index config is passed.